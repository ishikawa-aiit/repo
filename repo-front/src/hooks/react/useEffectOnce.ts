import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
