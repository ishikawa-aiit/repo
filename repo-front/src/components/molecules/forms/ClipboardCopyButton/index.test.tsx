import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { ClipboardCopyButton } from '.'

describe('snapshot', () => {
  it('正常にレンダリングされている', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    const { container } = render(<ClipboardCopyButton copyText="" onCopied={() => {}} />)

    expect(container).toMatchSnapshot()
  })
})

describe('copy', () => {
  it('onCopied の処理が呼ばれる', async () => {
    const handleOnCopy = jest.fn<() => void, never[]>()

    render(<ClipboardCopyButton copyText="" onCopied={handleOnCopy} />)

    fireEvent.click(screen.getByTestId('ClipboardCopyButton'))

    await waitFor(() => {
      expect(handleOnCopy).toBeCalled()
    })
  })
})
