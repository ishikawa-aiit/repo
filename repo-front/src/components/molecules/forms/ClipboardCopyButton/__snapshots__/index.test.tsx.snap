// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.emotion-1 {
  width: 5px;
}

<div>
  <button
    class="chakra-button emotion-0"
    data-cy="ClipboardCopyButton"
    data-testid="ClipboardCopyButton"
    type="button"
  >
    <span
      class="chakra-text emotion-1"
    >
      <svg
        aria-hidden="true"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </span>
  </button>
</div>
`;
